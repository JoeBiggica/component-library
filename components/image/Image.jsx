import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import styles from './Image.scss';

const FillMode = {
	EXACT: 'exact',
	FIT: 'fit',
	COVER: 'cover',
};

const Placeholder = {
	LOGO_LIGHT: 'logo-light',
	LOGO_DARK: 'logo-dark',
};

class Image extends PureComponent {
	static propTypes = {
		className: PropTypes.string,
		style: PropTypes.object,
		src: PropTypes.string,
		is_amp: PropTypes.bool,
		alt: PropTypes.string,
		aspect_ratio: PropTypes.number,
		fill_mode: PropTypes.string,
		placeholder: PropTypes.oneOf([false, ...Object.values(Placeholder)]),
		forwarded_ref: PropTypes.object,
	};

	static defaultProps = {
		fill_mode: FillMode.EXACT,
		placeholder: Placeholder.LOGO_LIGHT,
		is_amp: false,
	};

	static FillMode = FillMode;

	static Placeholder = Placeholder;

	constructor(props) {
		super(props);

		this.state = {
			loaded: false,
			src: this.props.src,
		};
	}

	static getDerivedStateFromProps(props, state) {
		if (props.src !== state.src) {
			return {
				src: props.src,
				loaded: false,
			};
		}

		return null;
	}

	handleImageLoad = () => {
		this.setState({
			loaded: true,
		});
	};

	render() {
		const {
			className,
			style,
			alt,
			aspect_ratio,
			fill_mode,
			placeholder,
			forwarded_ref,
			is_amp,
		} = this.props;

		const { src, loaded } = this.state;

		const container_style = aspect_ratio
			? {
					height: 0,
					paddingBottom: `${(1 / aspect_ratio) * 100}%`,
			  }
			: undefined;

		return (
			<div className={className} style={style} ref={forwarded_ref}>
				<div className={styles['size-container']}>
					<div
						className={classnames(
							styles['image-container'],
							aspect_ratio && styles['lock-ratio'],
							placeholder && styles[placeholder],
							loaded && styles['loaded']
						)}
						style={container_style}
					>
						{is_amp
							? src && (
									<amp-img
										key={src}
										className={classnames(styles['image'], styles[fill_mode])}
										alt={alt}
										src={src}
										title={alt}
										onLoad={this.handleImageLoad}
										layout='fill'
									/>
							  )
							: src && (
									<img
										key={src}
										className={classnames(styles['image'], styles[fill_mode])}
										src={src}
										alt={alt}
										title={alt}
										onLoad={this.handleImageLoad}
									/>
							  )}
					</div>
				</div>
			</div>
		);
	}
}

export default Image;
