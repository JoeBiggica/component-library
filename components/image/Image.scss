@import 'styles/_colors.scss';
@import 'styles/_eases.scss';

@keyframes fadein {
    from {
    	opacity: 0;
    }
    to {
    	opacity: 1;
    }
}

.size-container {
	margin: 0 auto;
	height: inherit;
}

.image-container {
	position: relative;
	height: inherit;
	z-index: 0;

	&.logo-light:after,
	&.logo-dark:after {
		content: '';
		display: block;
		position: absolute;
		z-index: 0;
		top: 0;
		left: 0;
		width: 100%;
		height: 100%;
		background-repeat: no-repeat;
		background-size: auto 35%;
		background-position: center center;
		opacity: .1;
	}

	&.logo-light {
		background-color: $background-grey;

		&:after {
			background-image: url(./five-rings-black.svg);
		}
	}

	&.logo-dark {
		background-color: $almost-black;
		
		&:after {
			background-image: url(./five-rings-white.svg);
		}
	}
}

.image-container:not(.loaded) .image {
	display: none;
}

.image {
	position: relative;
	z-index: 1;
	display: block;
	width: 100%;
	height: 100%;
	animation: fadein 0.25s $principleDefaultEase both;

	&.fit {
		object-fit: contain;
	}

	&.cover {
		object-fit: cover;
	}
}

.lock-ratio .image {
	position: absolute;
	height: 100%;
}
