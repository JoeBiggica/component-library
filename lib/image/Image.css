/*  imported from Image.scss  */

@keyframes Image__fadein___agl3V {
  from {
    opacity: 0; }
  to {
    opacity: 1; } }

.Image__size-container___1F01j {
  margin: 0 auto;
  height: inherit; }

.Image__image-container___3XUvb {
  position: relative;
  height: inherit;
  z-index: 0; }
  .Image__image-container___3XUvb.Image__logo-light___1o9eH:after, .Image__image-container___3XUvb.Image__logo-dark___gBP8T:after {
    content: '';
    display: block;
    position: absolute;
    z-index: 0;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    background-repeat: no-repeat;
    background-size: auto 35%;
    background-position: center center;
    opacity: .1; }
  .Image__image-container___3XUvb.Image__logo-light___1o9eH {
    background-color: #e6e6e6; }
    .Image__image-container___3XUvb.Image__logo-light___1o9eH:after {
      background-image: url(./five-rings-black.svg); }
  .Image__image-container___3XUvb.Image__logo-dark___gBP8T {
    background-color: #313132; }
    .Image__image-container___3XUvb.Image__logo-dark___gBP8T:after {
      background-image: url(./five-rings-white.svg); }

.Image__image-container___3XUvb:not(.Image__loaded___2w56Y) .Image__image___6wPci {
  display: none; }

.Image__image___6wPci {
  position: relative;
  z-index: 1;
  display: block;
  width: 100%;
  height: 100%;
  animation: Image__fadein___agl3V 0.25s cubic-bezier(0.25, 0.1, 0.25, 1) both; }
  .Image__image___6wPci.Image__fit___1uHCx {
    object-fit: contain; }
  .Image__image___6wPci.Image__cover___2NcXy {
    object-fit: cover; }

.Image__lock-ratio___s5qTP .Image__image___6wPci {
  position: absolute;
  height: 100%; }
