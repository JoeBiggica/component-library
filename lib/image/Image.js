import "core-js/modules/es6.string.iterator";
import "core-js/modules/es6.array.from";
import "core-js/modules/es6.regexp.to-string";
import "core-js/modules/web.dom.iterable";
import "core-js/modules/es6.array.iterator";
import "core-js/modules/es6.object.to-string";
import "core-js/modules/es7.object.values";
import "core-js/modules/es7.symbol.async-iterator";
import "core-js/modules/es6.symbol";
import "core-js/modules/es6.object.set-prototype-of";

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance"); }

function _iterableToArray(iter) { if (Symbol.iterator in Object(iter) || Object.prototype.toString.call(iter) === "[object Arguments]") return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = new Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

require("./Image.css");

var styles = {
  "size-container": "Image__size-container___1F01j",
  "image-container": "Image__image-container___3XUvb",
  "logo-light": "Image__logo-light___1o9eH",
  "logo-dark": "Image__logo-dark___gBP8T",
  "loaded": "Image__loaded___2w56Y",
  "image": "Image__image___6wPci",
  "fadein": "Image__fadein___agl3V",
  "fit": "Image__fit___1uHCx",
  "cover": "Image__cover___2NcXy",
  "lock-ratio": "Image__lock-ratio___s5qTP"
};
var FillMode = {
  EXACT: 'exact',
  FIT: 'fit',
  COVER: 'cover'
};
var Placeholder = {
  LOGO_LIGHT: 'logo-light',
  LOGO_DARK: 'logo-dark'
};

var Image =
/*#__PURE__*/
function (_PureComponent) {
  _inherits(Image, _PureComponent);

  function Image(props) {
    var _this;

    _classCallCheck(this, Image);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(Image).call(this, props));

    _defineProperty(_assertThisInitialized(_this), "handleImageLoad", function () {
      _this.setState({
        loaded: true
      });
    });

    _this.state = {
      loaded: false,
      src: _this.props.src
    };
    return _this;
  }

  _createClass(Image, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          className = _this$props.className,
          style = _this$props.style,
          alt = _this$props.alt,
          aspect_ratio = _this$props.aspect_ratio,
          fill_mode = _this$props.fill_mode,
          placeholder = _this$props.placeholder,
          forwarded_ref = _this$props.forwarded_ref,
          is_amp = _this$props.is_amp;
      var _this$state = this.state,
          src = _this$state.src,
          loaded = _this$state.loaded;
      var container_style = aspect_ratio ? {
        height: 0,
        paddingBottom: "".concat(1 / aspect_ratio * 100, "%")
      } : undefined;
      return React.createElement("div", {
        className: className,
        style: style,
        ref: forwarded_ref
      }, React.createElement("div", {
        className: styles['size-container']
      }, React.createElement("div", {
        className: classnames(styles['image-container'], aspect_ratio && styles['lock-ratio'], placeholder && styles[placeholder], loaded && styles['loaded']),
        style: container_style
      }, is_amp ? src && React.createElement("amp-img", {
        key: src,
        className: classnames(styles['image'], styles[fill_mode]),
        alt: alt,
        src: src,
        title: alt,
        onLoad: this.handleImageLoad,
        layout: "fill"
      }) : src && React.createElement("img", {
        key: src,
        className: classnames(styles['image'], styles[fill_mode]),
        src: src,
        alt: alt,
        title: alt,
        onLoad: this.handleImageLoad
      }))));
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(props, state) {
      if (props.src !== state.src) {
        return {
          src: props.src,
          loaded: false
        };
      }

      return null;
    }
  }]);

  return Image;
}(PureComponent);

_defineProperty(Image, "propTypes", {
  className: PropTypes.string,
  style: PropTypes.object,
  src: PropTypes.string,
  is_amp: PropTypes.bool,
  alt: PropTypes.string,
  aspect_ratio: PropTypes.number,
  fill_mode: PropTypes.string,
  placeholder: PropTypes.oneOf([false].concat(_toConsumableArray(Object.values(Placeholder)))),
  forwarded_ref: PropTypes.object
});

_defineProperty(Image, "defaultProps", {
  fill_mode: FillMode.EXACT,
  placeholder: Placeholder.LOGO_LIGHT,
  is_amp: false
});

_defineProperty(Image, "FillMode", FillMode);

_defineProperty(Image, "Placeholder", Placeholder);

export default Image;
//# sourceMappingURL=Image.js.map