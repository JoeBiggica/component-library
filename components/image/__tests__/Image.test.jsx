import React from 'react';
import { render } from '@testing-library/react';

import Image from '../Image';

const commonProps = {
	is_amp: false,
	src: 'source_path',
	alt: 'alt',
	fill_mode: 'fit',
	placeholder: 'logo-light',
};

describe('<Image />', () => {
	it('should render Image component correctly', () => {
		const { container } = render(<Image {...commonProps} />);

		expect(container).toMatchSnapshot();
	});
	it('should render Image component with amp tag', () => {
		const { container } = render(<Image {...commonProps} is_amp />);

		expect(container).toMatchSnapshot();
	});

	it('should render Image component with aspect_ratio', () => {
		const { container } = render(
			<Image {...commonProps} aspect_ratio={380 / 249} />
		);

		expect(container).toMatchSnapshot();
	});
});
