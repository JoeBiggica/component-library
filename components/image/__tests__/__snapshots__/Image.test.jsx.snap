// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Image /> should render Image component correctly 1`] = `
<div>
  <div>
    <div
      class="size-container"
    >
      <div
        class="image-container logo-light"
      >
        <img
          alt="alt"
          class="image fit"
          src="source_path"
          title="alt"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<Image /> should render Image component with amp tag 1`] = `
<div>
  <div>
    <div
      class="size-container"
    >
      <div
        class="image-container logo-light"
      >
        <amp-img
          alt="alt"
          classname="image fit"
          layout="fill"
          src="source_path"
          title="alt"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<Image /> should render Image component with aspect_ratio 1`] = `
<div>
  <div>
    <div
      class="size-container"
    >
      <div
        class="image-container lock-ratio logo-light"
        style="height: 0px; padding-bottom: 65.52631578947367%;"
      >
        <img
          alt="alt"
          class="image fit"
          src="source_path"
          title="alt"
        />
      </div>
    </div>
  </div>
</div>
`;
